### PR TITLE
txn: Fix to the prewrite requests retry problem by using TxnStatusCache

### PR DESCRIPTION
**Pull request by MyonKeminta**
### What is changed and how it works?

Issue Number: Ref #11187

What's Changed:

```commit-message
```

### Related changes

- Requires:
  - https://github.com/tikv/tikv/pull/15747
  - https://github.com/tikv/tikv/pull/15749

- PR to update `pingcap/docs`/`pingcap/docs-cn`:

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
Fix the problem that in environments that network is very unstable, if a prewrite request is retried ant the TiKV recieves the retried request before the original request, it's possible to lead to transactoin correctness issue.
```

### CODE REVIEW

**Comment by MyonKeminta** [comment]
This PR becames too large.
Changes to lru and adding TxnStatusCache module will be separated out as new PRs.

File: src/storage/txn/scheduler.rs [hunk 1]
```diff
@@ -484,6 +487,8 @@ impl<E: Engine, L: LockManager> TxnScheduler<E, L> {
             quota_limiter,
             resource_manager,
             feature_gate,
+            // TODO: Make it configurable
+            txn_status_cache: TxnStatusCache::new(40 * 1000 * 128),
```
Comment (ekexium) [inline src/storage/txn/scheduler.rs:491]: Can we explain how this is decided?

**Review by ekexium — APPROVED** [review]

File: src/storage/txn/scheduler.rs (hunk 1 above)
Comment (MyonKeminta) [reply src/storage/txn/scheduler.rs:491]: Sorry, I just noticed that some changes seem to be lost from this PR.. I'll update it...

File: src/storage/txn/scheduler.rs [hunk 2]
```diff
@@ -882,6 +887,13 @@ impl<E: Engine, L: LockManager> TxnScheduler<E, L> {
 
         self.on_acquired_locks_finished(metadata, new_acquired_locks);
 
+        if !known_txn_status.is_empty() {
```
Comment (cfzjywxk) [inline src/storage/txn/scheduler.rs:890]: The `result` should be checked here as the cache should be updated only when the `async-write` finishes successfully. Unit-test to cover this path is also needed.

File: src/storage/config.rs [hunk 3]
```diff
@@ -31,6 +31,11 @@ const DEFAULT_SCHED_PENDING_WRITE_MB: u64 = 100;
 const DEFAULT_RESERVED_SPACE_GB: u64 = 5;
 const DEFAULT_RESERVED_RAFT_SPACE_GB: u64 = 1;
 
+// In tests, we've observed 1.2M entries in the TxnStatusCache. We
+// conservatively set the limit to 5M entries in total.
+// As TxnStatusCache have 128 slots by default. We round it to 5.12M.
+const DEFAULT_TXN_STATUS_CACHE_CAPACITY: usize = 40_000 * 128;
```
Comment (cfzjywxk) [inline src/storage/config.rs:37]: Better to comment how much memory would be used in bytes approximately here.

File: src/storage/mod.rs [hunk 4]
```diff
@@ -10869,4 +10871,326 @@ mod tests {
...
+            .unwrap();
+        rx.recv().unwrap();
+        assert_eq!(
+            storage
+                .sched
+                .get_txn_status_cache()
+                .get_no_promote(90.into())
+                .unwrap(),
+            100.into()
+        );
```
Comment (cfzjywxk) [inline src/storage/mod.rs:11194]: `check_txn_status` and `check_secondary_locks` are not coverred here?

**Comment by cfzjywxk** [comment]
Is our next step is to add txn cache read usages for other possible pathes? For example checking results from cache first for `check_txn_status` and `resolve_lock`.

**Comment by MyonKeminta** [comment]
> Is our next step is to add txn cache read usages for other possible pathes? For example checking results from cache first for `check_txn_status` and `resolve_lock`.

Making use of it to speed up resolving lock procedure might not be a simple task. I'm wondering if this is indeed helpful for our large transaction project.

**Review by cfzjywxk — APPROVED** [review]
Rest LGTM

**Pull request merged**